### PR TITLE
fix: fix checkout root logic to work inside build_engine

### PR DESCRIPTION
**Pull request by eseidel**
The previous logic assumed cutler was in shorebird.

I also tested the checkout root finding logic.

### CODE REVIEW

**Comment by eseidel** [comment]
I'm not entirely sure why it's failing.  It seems to be using the "cutler" flag when uploading, but CodeCov also isn't showing data for the "cutler" flag?

Is it trying to enforce 50% coverage per-file, or overall?

I can look again monday.  Regardless of the coverage issues I think this is a good change. :). FYI @felangel

**Comment by eseidel** [comment]
codecov claims there is no data on `cutler` as a flag:
https://app.codecov.io/gh/shorebirdtech/build_engine/flags

I checked and it appears to be using the "cutler" flag (even though it's not in a packages/cutler directory and our action.yaml logic could get confused by that).

**Review by felangel — APPROVED** [review]

File: cutler/lib/cutler.dart [hunk 1]
```diff
@@ -35,43 +36,14 @@ class Cutler extends CommandRunner<int> {
...
     final results = super.parse(args);
 
     final rootDir = results['root'] as String? ?? fallbackRootDir();
 
     final missingDirs = missingDirectories(rootDir);
-    if (missingDirs.isNotEmpty) {
+    if (rootDir == null || missingDirs.isNotEmpty) {
       logger
         ..err('Could not find a valid checkouts root.')
         ..err('--root must be a directory containing the '
```
Comment (bryanoltman) [inline cutler/lib/cutler.dart:49]: Consider using multiline strings here instead of `\n`

File: cutler/lib/cutler.dart [hunk 2]
```diff
@@ -94,3 +66,37 @@ class Cutler extends CommandRunner<int> {
     return results;
   }
 }
+
+/// Lists directories missing from the proposed root directory.
+Iterable<String> missingDirectories(@visibleForTesting String? rootDir) {
```
Comment (bryanoltman) [inline cutler/lib/cutler.dart:71]: Which directories do we expect to have (i.e., what is the full set?)

Also, I don't think I've ever seen `@visibleForTesting` used on a parameter before. How is annotating the parameter different than annotating the function?

File: cutler/lib/cutler.dart [hunk 3]
```diff
@@ -94,3 +66,37 @@ class Cutler extends CommandRunner<int> {
...
+    p.dirname(buildEngineDir),
+  ];
+  for (final directory in fallbackDirectories) {
+    if (missingDirectories(directory).isEmpty) {
+      logger.info('Using $directory as checkouts root.');
+      return directory;
+    }
+  }
+  logger.err('Failed to find a valid checkouts root, tried:\n'
+      '${fallbackDirectories.join('\n')}');
```
Comment (bryanoltman) [inline cutler/lib/cutler.dart:99]: Same here re: multiline strings

**Review by bryanoltman — APPROVED** [review]

File: cutler/lib/cutler.dart (hunk 1 above)
Comment (eseidel) [reply cutler/lib/cutler.dart:49]: Totally cheating.  We must work around the language lack of features... I mean avoid using the features the language has.  It's part of the challange. 🙃

File: cutler/lib/cutler.dart (hunk 2 above)
Comment (eseidel) [reply cutler/lib/cutler.dart:71]: Repo.values.map((repo) => repo.path) is the full set.

My thought was that `@visibleForTesting` would warn if someone tries to pass that parameter from non-test code?

File: cutler/lib/cutler.dart (hunk 3 above)
Comment (eseidel) [reply cutler/lib/cutler.dart:99]: A later refactor, but thank you for the tip. :)

**Pull request merged**